Make Figaro values more consistent

jessieay opened the pull request:
**What**:
  Use "true" and "false" instead of "yes" and "no"
  Raise RuntimeError if yes or no pop up

**Why**:
  Consistency ensures we are setting up features correctly
  Closes https://github.com/18F/identity-private/issues/851

jessieay commented:
I would vote to ignore the current code climate errors

1. `FigaroYamlParser#parse refers to value more than self (maybe move it to another class?)`
    
  This is the result of type-checking on whether what we are looking at is a Hash or a String. We want to know if something is a Hash because if it is we want to recursively loop through the values to check for "false" or "no'

2. Unused block argument - key. If it's necessary, use _ or _key as an argument name to indicate that it won't be used.

  It is used, but it is interpolated in `%q()` -- this is a case where I think Code Climate's regex     isn't smart enough

3.  `Do not use :: for method calls.`

  Unfortunately, we do not own `YAML::load` so I don't think we can do much there.

jessieay commented:
@dogweather I am ok with standardizing on that if others are on board. I tend to feel more comfortable with "true" and "false" in code-land but I see your point

dogweather commented:
I think it's fine to use 'true' and 'false' because it's such a standard. I.e., Rails uses them in its config.

jessieay commented:
> Hey @jessieay I'm taking a guess what the yaml parser class is for, but I'm not sure. It seems to enable nested yaml configuration? Also, the tests for it don't seem to actually test its functionality; just a particular response to bad input. ?

This tests that `config/application.yml` does not contain any values that are "yes" or "no" -- this file contains nested yml. It is copied over from https://github.com/18F/identity-idp/blob/master/config/application.yml.example on L26 of `bin/setup` so you can see what it looks like there.

All that we want this parser to do is raise when there is bad input. Perhaps we should rename the parser class if the functionality isnt making sense to you? (I can see how the name would make it confusing but if you read the GH issue related to this the desired functionality is pretty straightforward)

dogweather commented:
Ah hah - yeah, the names `*Parser` and `parse` threw me off. I think `*Validator` and `validate` describe better what it's doing - what do you think?

Also – it looks like this code will fail in production because it will try to grammar check the config file ... which we don't keep in the repo, and isn't used in production in the standard Figaro set up.

monfresh commented:
What Rubocop means by "do not use :: for method calls" is that the dot syntax is preferred: `YAML.load` versus `YAML::load`. 

As for the other offenses, they can be fixed by using `YAML.load_file`, which gives you a hash to work with. 

Also, there two bugs with your current code. You can test it locally by setting any key in `application.yml` to a value of `'yes'`, then running `spring stop`, then `rails c`.

- You will see that the error message that is raised does not interpolate the key and value, it says `#{key}` and `#{value}` literally: `You set #{key} to #{value} in config/application.yml. Please change it to true or false`. Since the error message is important to us, the test should make sure the message is correct, as opposed to only checking that it raises a `RuntimeError`.

- Using raise inside a loop will not raise for each failed case. It will only raise once at the first case, but we want to know all the cases that need to be fixed.

Here is how I would rewrite this:

```ruby
class FigaroYamlParser
  FILE = "#{Rails.root}/config/application.yml".freeze
  FIGARO_YAML = YAML.load_file(FILE).fetch(Rails.env, {})

  def parse(yaml_data = FIGARO_YAML)
    result = yaml_data.select { |_, value| %w(yes no).include?(value) }

    raise "These entries in config/application.yml have 'yes' or 'no' values. " \
          "Please change them to 'true' or 'false': #{result}"
  end
end
```

monfresh commented:
@dogweather `config/application.yml` is uploaded to the servers via [Capistrano](https://github.com/18F/identity-idp/blob/f18f1872fbd481c59a2f45e56b2f46657eb98863/config/deploy.rb#L13).

jessieay commented:
I updated this in response to your feedback, @monfresh and @dogweather 

- Renamed class to `FigaroYamlValidator` and renamed method to `validate` (instead of `parse`)
- Now using dot syntax instead of `::`
- Using `<<~HEREDOC` instead of `%q` so that interpolation works and there isn't weird indenting (was having issues with extra spaces with `%Q`) My first time using the squiggly heredoc! https://infinum.co/the-capsized-eight/articles/multiline-strings-ruby-2-3-0-the-squiggly-heredoc
- Updated logic / specs to have output for all bad values, not just first one encountered
- Updated spec to look for specific error message so we don't run into that interpolation rpoblem

monfresh commented on lib/figaro_yaml_validator.rb:
```diff
@@ -0,0 +1,30 @@
+      config/application.yml. Please change them to true or false
+    HEREDOC
+  end
+end
```
Was there something wrong with my solution? This seems more complicated than it needs to be. Also, this looks at all the values in all the Rails.env sections, which can cause duplicates. For example, if both the development and production entries for a key have an incorrect value, the message will look like this:

```
You have invalid values for allow_third_party_auth and allow_third_party_auth in config/application.yml. Please change them to true or false
```
Similarly, if the production value is incorrect, but the development value is correct, raising an error for the production entry when running the app in development might cause confusion, which is why I added `.fetch(Rails.env)` in my solution so that the validation only runs against the current environment.

By using `.load_file`, you get a Hash right away and you don't have to do any type checks.

The one thing I missed in my solution was an `if result` for the raise so that it only raises if the validation fails.

jessieay replied:
I like the idea of checking the whole file rather than just the env-specific stuff because in ensures we aren't going to use invalid values in any env. Does that make sense?

I see what you are saying about the error message not being specific to env, but I think someone should pretty easily be able to grep and find the right key?

Because I am checking all envs, I also need to deal with a nested hash, which is why this is more complicated.

monfresh replied:
I just realized that we have top-level entries that aren't part of a specific Rails.env entry, which is the reason we need to do the type checking and that we can't rely on fetching the Rails env alone.

Try this to get rid of the reek offense:

```ruby
elsif %w(yes no).include?(value)
```

And maybe add `.uniq` to bad_keys in the warning to remove dupes from the error message?

jessieay replied:
good one! I thought the error was related to type-checking on Hash but looks like this fixed reek

jessieay replied:
also added `uniq`

jessieay commented:
@monfresh @dogweather care to re-review / merge?

monfresh commented:
@dogweather must approve the changes before we can merge.

jessieay commented:
thanks for the review! @dogweather lmk if you'd like more changes before merging
